File: app/internal/models/external/subscription.go
package external

import (
	"FinancialTracker/internal/models"
)

// SubscriptionPayload is the client-facing subscription and billing summary.
type SubscriptionPayload struct {
	Tier                  string                        `json:"tier"`
	Billing               models.BillingPeriod          `json:"billing"`
	Limits                models.PlaidLimits            `json:"limits"`
	Plans                 []models.TierPlan             `json:"plans"`
	StripeConfigured      bool                          `json:"stripe_configured"`
	BillingEnabled        bool                          `json:"billing_enabled"`
	SubscriptionsEnabled  bool                          `json:"subscriptions_enabled"`
	HasActiveSubscription bool                          `json:"has_active_subscription"`
	CanChangePlan         bool                          `json:"can_change_plan"`
	Privileges            models.SubscriptionPrivileges `json:"privileges"`
}

// CheckoutSessionRequest starts a Stripe Checkout flow for a paid tier.
type CheckoutSessionRequest struct {
	Tier string `json:"tier"`
}

// CheckoutSessionResponse returns the hosted Stripe Checkout URL.
type CheckoutSessionResponse struct {
	URL string `json:"url"`
}

// BillingPortalResponse returns the hosted Stripe Customer Portal URL.
type BillingPortalResponse struct {
	URL string `json:"url"`
}

// ChangeSubscriptionRequest switches the user's subscription tier.
type ChangeSubscriptionRequest struct {
	Tier string `json:"tier"`
}

// ChangeSubscriptionResponse confirms a tier change.
type ChangeSubscriptionResponse struct {
	Subscription SubscriptionPayload `json:"subscription"`
}

File: frontend/src/lib/subscription.ts
import { clientApiRequest } from "./api";
import type {
  BillingPortalResponse,
  ChangeSubscriptionResponse,
  CheckoutSessionResponse,
  SubscriptionPayload,
} from "./types";

// getSubscription loads the user's plan and billing period.
export async function getSubscription(): Promise<SubscriptionPayload> {
  return clientApiRequest<SubscriptionPayload>("/api/v1/subscription");
}

// changeSubscription applies a skeleton tier change (development only).
export async function changeSubscription(tier: string): Promise<ChangeSubscriptionResponse> {
  return clientApiRequest<ChangeSubscriptionResponse>("/api/v1/subscription/change", {
    method: "POST",
    body: { tier },
  });
}

// createCheckoutSession starts Stripe Checkout for a paid tier.
export async function createCheckoutSession(tier: string): Promise<CheckoutSessionResponse> {
  return clientApiRequest<CheckoutSessionResponse>("/api/v1/subscription/checkout", {
    method: "POST",
    body: { tier },
  });
}

// createBillingPortal opens the Stripe Customer Portal for the signed-in user.
export async function createBillingPortal(): Promise<BillingPortalResponse> {
  return clientApiRequest<BillingPortalResponse>("/api/v1/subscription/portal", {
    method: "POST",
  });
}
